motionplan/tpSpaceRRT.go

```diff
@@ -0,0 +1,436 @@
+	trajK := curPtg.Trajectory(goalK)
+	pass := true
+	var lastNode *tpspace.TrajNode
+	var lastPose spatialmath.Pose
+
+	sinceLastCollideCheck := 0.
+	lastDist := 0.
+
+	// Check each point along the trajectory to confirm constraints are met
+	for _, trajPt := range trajK {
+		if trajPt.Dist > goalD {
+			// After we've passed randD, no need to keep checking, just add to RRT tree
+			break
+		}
+		sinceLastCollideCheck += (trajPt.Dist - lastDist)
+		trajState := &State{Position: spatialmath.Compose(nearest.Pose(), trajPt.Pose), Frame: mp.frame}
+		if sinceLastCollideCheck > mp.planOpts.Resolution {
+			ok, _ := mp.planOpts.CheckStateConstraints(trajState)
+			if !ok {
+				pass = false
+				break
+			}
+		}
+
+		lastPose = trajState.Position
+		lastNode = trajPt
+		lastDist = lastNode.Dist
+	}
+	if pass {
+		// add the last node in trajectory
+		successNode = newConfigurationCostNode(
+			referenceframe.FloatsToInputs([]float64{float64(ptgNum), float64(goalK), lastNode.Dist}),
+			nearest.Cost()+lastNode.Dist,
+			lastPose,
+		)
+	} else {
+		return nil
+	}
+
+	cand := &candidate{dist: bestDist, treeNode: nearest, newNode: successNode}
+
+	if !acceptableGoal && successNode != nil {
+		// check if this  successNode is too close to nodes already in the tree, and if so, do not add.
+
+		// Get nearest neighbor to new node that's already in the tree
+		nearest := nm.nearestNeighbor(ctx, mp.planOpts, successNode, rrt.maps.startMap)
+		dist := math.Inf(1)
+		if nearest != nil {
+			dist = mp.planOpts.DistanceFunc(&Segment{StartPosition: successNode.Pose(), EndPosition: nearest.Pose()})
+		}
+
+		// Ensure successNode is sufficiently far from the nearest node already existing in the tree
+		// If too close, don't add a new node
+		if dist < defaultIdenticalNodeDistance {
+			cand = nil
+		}
+	}
+	return cand
+}
+
+// extendMap grows the rrt map to the best candidate node if it is valid to do so.
+func (mp *tpspaceRRTMotionPlanner) extendMap(
+	ctx context.Context,
+	candidateNodes map[float64][2]node,
+	rrt *rrtParallelPlannerShared,
+	tpFrame tpspace.PTGProvider,
+) {
+	var addedNode node
+	// If we found any valid nodes that we can extend to, find the very best one and add that to the tree
+	if len(candidateNodes) > 0 {
+		bestDist := math.Inf(1)
+		var bestNode [2]node
+		for k, v := range candidateNodes {
+			if k < bestDist {
+				bestNode = v
+				bestDist = k
+			}
+		}
+
+		ptgNum := int(bestNode[1].Q()[0].Value)
+		randK := uint(bestNode[1].Q()[1].Value)
+
+		trajK := tpFrame.PTGs()[ptgNum].Trajectory(randK)
+
+		lastDist := 0.
+		sinceLastNode := 0.
+
+		for _, trajPt := range trajK {
+			if ctx.Err() != nil {
+				return
+			}
+			if trajPt.Dist > bestNode[1].Q()[2].Value {
+				// After we've passed goalD, no need to keep checking, just add to RRT tree
+				break
+			}
+			trajState := &State{Position: spatialmath.Compose(bestNode[0].Pose(), trajPt.Pose)}
+			sinceLastNode += (trajPt.Dist - lastDist)
```

Comment:
This since last node concept is used a couple times.  Wondering if there is common code we can pull out of both these places

Reply:
IMO this is simple enough that it's not worthwhile